BeepBoopHQ/beepboop-js
# Initial implementation of beepboop resourcer

#### What's this PR do?

Adds ability to connect to a Beep Boop Resourcer server and listen for events. Events should serve to notify a bot maker when users (team owners typically) have requested to add, update, remove the bot from their team.

This is the lowest level module for running on beep boop. A beepboop-botkit module will wrap this module and handle adding, updating, and removing teams the botkit way.

The most useful events are `message.add_resource`, `message.update_resource`, `message.remove_resource`. -- the namespaces are up for debate and worth discussing via this PR.
#### Where should the reviewer start?

It's pretty much all in the lib/resourcer.js. Worth looking at lib/retry.js as well.

Also, would be good to sanity check the README to make sure it's intuitive.
#### How should this be manually tested?

Against real beepboop resource server would be ideal.

set:
      export BEEPBOOP_TOKEN={the resourcer token},
      export BEEPBOOP_ID={the pod id}
      export BEEPBOOP_RESOURCER={the resourcer web socket url ex: 'ws://localhost:9000/ws'}

`node ./examples/simple.js`

trigger add, update, remove events on resourcer. you should see the console logs coming from the bot process.
#### Were the tests updated? yes but need coverage around retry.
